File: python/day5.py
#!/usr/local/bin/python3

import re

with open('../day5_input.txt') as f:
    lines = [l.rstrip() for l in f.readlines()]


def is_nice(string):
    bad_strings = ['ab', 'cd', 'pq', 'xy']
    no_bad_strings = not any(bad for bad in bad_strings if bad in string)

    vowels = set('aeiou')
    enough_vowels = len([c for c in string if c in vowels]) >= 3

    has_double_char = re.search(r'(.)\1', string) is not None

    return no_bad_strings and enough_vowels and has_double_char


nice_strings = list(filter(is_nice, lines))
print('Number of nice strings:', len(nice_strings))


# Part Two


def new_nice(string):
    has_repeated_pair = re.search(r'(.{2}).*\1', string) is not None
    has_repeat_one_apart = re.search(r'(.).\1', string) is not None
    return has_repeated_pair and has_repeat_one_apart


new_nice_strings = list(filter(new_nice, lines))
print('Number of new nice strings:', len(new_nice_strings))

File: javascript/day5.js
#!/usr/local/bin/node

'use strict';
var fs = require('fs');

var instructions = fs.readFileSync('../day5_input.txt', 'utf-8')
    .trimRight()
    .split('\n');


// Part one

var is_nice = function (string) {
    var vowels = new Set('aeiou');
    var string_array = Array.from(string.toLowerCase());
    var enough_vowels = string_array.filter(c => vowels.has(c)).length >= 3;

    var no_forbidden_strings = string.search(/ab|cd|pq|xy/) === -1;
    var double_character = string.search(/(.)\1/) !== -1;

    return enough_vowels && no_forbidden_strings && double_character;
};

console.log('Nice strings:', instructions.filter(is_nice).length);


// Part two

var new_nice = function (string) {
    var repeated_pair = string.search(/(.{2}).*\1/) !== -1;
    var repeated_one_apart = string.search(/(.).\1/) !== -1;
    return repeated_pair && repeated_one_apart;
};

console.log('New nice strings:', instructions.filter(new_nice).length);
